In [1]:
# Using this as an actual protein in the csv file to
# attempt to match the DSSP file in DataFraming.

In [2]:
# Importing Libraries
from biopandas.pdb import PandasPdb
import pandas as pd # About to change some options 
import numpy as np
# Prior to modifications 
# pd.options.display.max_rows
# Prints 60 as max rows
# pd.set_option("display.max_rows", None)
# print(pd.options.display.max_rows)

In [3]:
# Securely fetching the protein and displaying
ppdb = PandasPdb().fetch_pdb('1A2T')
ppdb.df['ATOM'].head()
df = pd.DataFrame(ppdb.df['ATOM'])

In [4]:
df.head()

record_name  atom_number blank_1 atom_name alt_loc residue_name blank_2  \
0        ATOM            1                 N                  LEU           
1        ATOM            2                CA                  LEU           
2        ATOM            3                 C                  LEU           
3        ATOM            4                 O                  LEU           
4        ATOM            5                CB                  LEU           

  chain_id  residue_number insertion  ... x_coord  y_coord  z_coord  \
0        A               7            ...  -0.503   37.453   21.068   
1        A               7            ...   0.788   36.714   21.079   
2        A               7            ...   1.794   37.628   20.396   
3        A               7            ...   1.407   38.624   19.799   
4        A               7            ...   0.639   35.386   20.331   

   occupancy  b_factor  blank_4 segment_id element_symbol charge  line_idx  
0        1.0     53.40                                  N    NaN       375  
1        1.0     48.40                                  C    NaN       376  
2        1.0     50.49                                  C    NaN       377  
3        1.0     54.51                                  O    NaN       378  
4        1.0     44.96                                  C    NaN       379  

[5 rows x 21 columns]

In [5]:
def organizeddf(df):
    df1 = df.copy()
    del_cols = ['blank_1', 'alt_loc', "blank_2", 'insertion'
               ,'blank_3', 'occupancy', 'x_coord', 'y_coord', 'z_coord'
               ,'blank_4', 'segment_id', 'charge', 'line_idx']
    sorted_cols = ['record_name', 'atom_name', 'element_symbol', 'residue_name', 'chain_id'
                  ,'residue_number','b_factor']
# This command successfully drops those listed columns
    df1.drop(labels = del_cols,inplace=True, axis=1)
    df1.set_index('atom_number', inplace=True)
    df1 = df1[sorted_cols]
    return df1

In [15]:
df1 = organizeddf(df)
df1.head(20)

record_name atom_name element_symbol residue_name chain_id  \
atom_number                                                              
1                  ATOM         N              N          LEU        A   
2                  ATOM        CA              C          LEU        A   
3                  ATOM         C              C          LEU        A   
4                  ATOM         O              O          LEU        A   
5                  ATOM        CB              C          LEU        A   
6                  ATOM        CG              C          LEU        A   
7                  ATOM       CD1              C          LEU        A   
8                  ATOM       CD2              C          LEU        A   
9                  ATOM         N              N          HIS        A   
10                 ATOM        CA              C          HIS        A   
11                 ATOM         C              C          HIS        A   
12                 ATOM         O              O          HIS        A   
13                 ATOM        CB              C          HIS        A   
14                 ATOM        CG              C          HIS        A   
15                 ATOM       ND1              N          HIS        A   
16                 ATOM       CD2              C          HIS        A   
17                 ATOM       CE1              C          HIS        A   
18                 ATOM       NE2              N          HIS        A   
19                 ATOM         H              H          HIS        A   
20                 ATOM       HD1              H          HIS        A   

             residue_number  b_factor  
atom_number                            
1                         7     53.40  
2                         7     48.40  
3                         7     50.49  
4                         7     54.51  
5                         7     44.96  
6                         7     39.75  
7                         7     36.61  
8                         7     37.86  
9                         8     50.58  
10                        8     49.59  
11                        8     45.25  
12                        8     43.75  
13                        8     57.60  
14                        8     70.89  
15                        8     77.91  
16                        8     77.90  
17                        8     80.45  
18                        8     81.34  
19                        8      0.00  
20                        8      0.00

In [7]:
path = '''/home/sanctus/Git/Proteomics-Research/data/csv
'''
df1a2t = df1.copy()
df1a2t.residue_number.fillna(value=('NaN'))
df1a2t.shape

(1337, 7)

In [16]:
# Returns a data conditioned DF
dfstd = organizeddf(df)
resbdf = dfstd.iloc[:,5:6]
resbdf['residue_number']
residues = resbdf['residue_number'].fillna(0)
print(residues)
#print(residues)
#print(type(residues))
# residues values as full numpy array 
residx = resbdf['residue_number'].unique()
print(residx)
# print(residues) prints all full values as an numpy array
# rescount = pd.Series(data = residues, )
# residues2 just gives me the length of the array of unique elements
# residues2 = resbdf['residue_number'].nunique()
resdfIdx = pd.Index(residx)
resdfval = pd.Series(residues)
#print(resdfIdx)
resdfcounts = resdfval.value_counts()
print(resdfcounts) # This prints out an unsorted index of unique values with the number of index rows it occupies.

atom_number
1         7
2         7
3         7
4         7
5         7
6         7
7         7
8         7
9         8
10        8
11        8
12        8
13        8
14        8
15        8
16        8
17        8
18        8
19        8
20        8
21        8
22        9
23        9
24        9
25        9
26        9
27        9
28        9
29        9
30        9
       ... 
1319    139
1320    139
1321    139
1322    139
1323    139
1324    139
1325    140
1326    140
1327    140
1328    140
1329    140
1330    140
1331    140
1332    140
1333    140
1334    140
1335    140
1336    140
1337    140
1338    140
1339    140
1340    140
1341    141
1342    141
1343    141
1344    141
1345    141
1346    141
1347    141
1348    141
Name: residue_number, Length: 1337, dtype: int64
[  7   8   9  10  11  12  13  14  15  16  17  18  19  20  21  22  24  25
  26  27  28  29  30  31  32  33  34  35  36  37  38  39  40  41  42  43
  44  45  46  47  48  49  50  51  52  53  54  55  56  57  58 

### Above shows the row and residue chain 

In [10]:
dfrescounts = pd.DataFrame(data=resdfcounts, index=resdfIdx)
dfrescounts.head()

residue_number
7                8
8               13
9               13
10              10
11               7

## Problem 1
### * Need to get the Average b_factor per residues
* List total number of unique items which should get  the total number of residues
* Do each of those have a set amount of do they differ? 

In [11]:
# Lets try a unique count of of the items in that specific column first
dfrescounts.cumsum().head()
# Cumulative sum of unique residue vaues equals the number of rows of the original pdb
# This indicates that each atom in the protein has a residue assignment. None that are not assigned.

residue_number
7                8
8               21
9               34
10              44
11              51

#### Total number of unique residues goes from 7 - 141, 134 total. 
* The residues have different amounts of atoms in them.
    * Need to find a way to grab those b factors.
        * Would be nice if it could be displayed in an array.
        * Not sure how to slice each set of differing residue numbers

In [25]:
# Using this DataFrame to show the b values.
resgroupbydf = dfstd.iloc[:,5:]
resgroupbydf.head(20)
# Creating a groupby object that groups by the different residue_number values as index with those appropriate
# b_factors
group = resgroupbydf['b_factor'].groupby(resgroupbydf['residue_number'])
# use b_factor values groups by residue number
group.groups
# Prints off the groupby object from Pandas and shows index rows.
# This explicitly specifies the rows of the DF to each residue unique value

{7: Int64Index([1, 2, 3, 4, 5, 6, 7, 8], dtype='int64', name='atom_number'),
 8: Int64Index([9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21], dtype='int64', name='atom_number'),
 9: Int64Index([22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34], dtype='int64', name='atom_number'),
 10: Int64Index([35, 36, 37, 38, 39, 40, 41, 42, 43, 44], dtype='int64', name='atom_number'),
 11: Int64Index([45, 46, 47, 48, 49, 50, 51], dtype='int64', name='atom_number'),
 12: Int64Index([52, 53, 54, 55, 56, 57], dtype='int64', name='atom_number'),
 13: Int64Index([58, 59, 60, 61, 62, 63, 64, 65, 66], dtype='int64', name='atom_number'),
 14: Int64Index([67, 68, 69, 70, 71, 72, 73, 74, 75], dtype='int64', name='atom_number'),
 15: Int64Index([76, 77, 78, 79, 80, 81, 82, 83, 84], dtype='int64', name='atom_number'),
 16: Int64Index([85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97], dtype='int64', name='atom_number'),
 17: Int64Index([98, 99, 100, 101, 102, 103], dtype='int64', name='atom_number'),
 1

### Applying operation on the group to obtain results for each individual residue number.

In [48]:
group.sum().head(10)
# Hand calculated res 7 and obtained 365.98 from the values.
# This groupby function seems to work appropriately.
print(group.mean().head(10))
# Mean gives appropriate answer as well.
print(type(group.mean()))
# Type is a Pandas Series.

residue_number
7     45.747500
8     48.866154
9     27.430769
10    37.422000
11    34.125714
12    26.063333
13    23.567778
14    28.446667
15    30.510000
16    30.486154
Name: b_factor, dtype: float64
<class 'pandas.core.series.Series'>


### Outputting for verifiable results, will optimize code in future.

In [39]:
# Using to_csv 
Avgbfac1a2t = group.mean()
df1a2tavgb = pd.DataFrame(Avgbfac1a2t)
df1a2tavgb.head()

b_factor
residue_number           
7               45.747500
8               48.866154
9               27.430769
10              37.422000
11              34.125714

In [49]:
df1a2tavgb.to_csv('Avgbfac1a2t.csv')
# Confirmed this file was there with ls. 
# File is readable!
df1a2tavgb.to_excel('Avgbfac1a2t.xlsx')
# Had to install openpyxl into the PP3 venv to get this to write to excel
# Excel file is created for this dict.

'1a2t DataFraming.ipynb'
 Avgbfac1a2t.csv
 Avgbfac1a2t.xlsx
'Biopandas Dataframing.ipynb'
'Biopandas Dataframing Jason.ipynb'
'Datamining PDB for Computations.ipynb'
'Reading and Extracting DSSP file information.ipynb'
